### PR TITLE
Document unsafety in core::slice::memchr

### PR DESCRIPTION
**Pull request by yokodake**
Contributes to #66219

Note sure if that's good enough, especially for the `align_to` call. 
The docs only mention transmuting and I don't think that everything related to reference lifetimes and state validity mentioned in the [nomicon](https://doc.rust-lang.org/nomicon/transmutes.html) are relevant here.

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
It looks like CI is failing here, but this looks good to me. I think you may have raced with https://github.com/rust-lang/rust/pull/76488, though. I'm happy to accept either, whichever one is ready first probably :)

**Comment by KodrAus** [comment]
Hi @yokodake :wave:

It looks like we've got a few failing tidy checks on this one:

```
tidy error: /checkout/library/core/src/slice/memchr.rs:118: undocumented unsafe
tidy error: /checkout/library/core/src/slice/memchr.rs:119: trailing whitespace
```

**Comment by yokodake** [comment]
Sorry didn't knew about these tidy checks. 
fixed it

**Comment by KodrAus** [comment]
It looks like we've refactored our `memchr` implementation a bit, but I think these comments will still be valid.

@yokodake would you like to rebase your PR?

**Comment by KodrAus** [comment]
Ah my bad! Looks like the rebase did happen 🙂 

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

**Comment by KodrAus** [comment]
This looks good to me! Thanks for working on it @yokodake!

@bors r+ rollup

**Comment by pietroalbini** [comment]
@bors r-

The PR failed but bors forgot about that during synchronize.

**Comment by yokodake** [comment]
What kind of test could have failed, I only added comments.
The only thing I can see in the logs is `completed with exit code 1`.
Do I have to do another rebase? There doesn't seem to be any merge conflict though.

**Comment by oliviacrain** [comment]
@yokodake: It seems like this build failure doesn't have anything to do with your PR- there was (is?) an intermittent issue with git on i686-mingw causing builds to fail (#73992). You might want to ping someone with bors privileges and get them to retry the build.

**Comment by yokodake** [comment]
ok, thanks.

@KodrAus 

@rustbot modify labels: -S-waiting-on-review +S-waiting-on-author

File: library/core/src/slice/memchr.rs [hunk 1]
```diff
@@ -105,6 +105,8 @@ pub fn memrchr(x: u8, text: &[u8]) -> Option<usize> {
     let (min_aligned_offset, max_aligned_offset) = {
         // We call this just to obtain the length of the prefix and suffix.
         // In the middle we always process two chunks at once.
+        // SAFETY: transmuting `[u8]` to `[usize]` is safe except for size differences
+        // which are handled by `align_to`.
```
Comment (sdroege) [inline library/core/src/slice/memchr.rs:109]: That's only safe if the `[u8]` is correctly aligned (and has a size that holds an integer multiple of `usize`s as you write here AFAIU). Might be worth mentioning that `align_to` also takes care of aligning properly (does it?).

Comment (Mark-Simulacrum) [reply library/core/src/slice/memchr.rs:109]: The more precise safety condition would be that `[u8; size_of::<usize>]` has the same validity invariants as usize, but I would not block on that in this PR.

**Pull request merged**